### PR TITLE
Unblock ProtoMPRS to control determinism of DataPipe in single/multi-processing and dist/non-dist env

### PR DESCRIPTION
**Pull request by ejguan**
This PR temporarily extend `PrototypingMultiProcessingReadingService` to fully control the determinism of the pipeline in the combinations of:
- Single/Multi-processing
- Distributed/Non-distributed
When we have `SequentialReadingService` ready to combine `DistributedReadingService` and `PrototypingMultiProcessingReadingService`, a few code should be removed. And, for in-process reading service, we still need a method to isolate global RNGs to prevent data-pipeline interferes randomness against model.

For multiprocessing case, it will set the same random seed for `Shuffler` and set different deterministic seeds for global RNGs including `python.random`, `torch` and `numpy` within each subprocess.
For distributed case, it will share the same random seed for `Shuffler` across all distributed subprocesses to guarantee the shuffle order before sharding.

Tests:
All tests are executed in the combinations of the above environments
- [x] Validate the same seed will generate the same order of data
- [x] Validate different seeds will generate different order of data
- [x] Validate the data after shuffle and sharding in each worker are mutually exclusive and collectively exhaustive with/without manual seed

There is one missing test I will add tmrw
- [x] Validate subprocess-local RNGs like `random`, `torch` and `numpy` are properly set with different seeds.

### CODE REVIEW

File: torchdata/dataloader2/communication/eventloop.py [hunk 1]
```diff
@@ -48,16 +48,16 @@ def DataPipeToQueuesLoop(source_datapipe, req_queue, res_queue, call_locally_fn=
...
         protocol_type(req_queue, res_queue),
         blocking_request_get=True,
-        reset_epoch_fn=call_on_reset_epoch,
+        reset_epoch_fn=call_on_epoch_reset,
     ):
         pass
 
 
-def SpawnProcessForDataPipeline(multiprocessing_ctx, datapipe, call_locally_fn=None, call_on_reset_epoch=None):
+def SpawnProcessForDataPipeline(multiprocessing_ctx, datapipe, call_on_process_init=None, call_on_epoch_reset=None):
```
Comment (ejguan) [inline torchdata/dataloader2/communication/eventloop.py:56]: I change these argument names to clarify the functionalities.

File: torchdata/dataloader2/reading_service.py [hunk 2]
```diff
@@ -174,38 +188,72 @@ def __init__(
         self.multiprocessing_context = multiprocessing_context
         self.processes = []
         self.datapipes = []
-        self.combined_datapipes = None
+        self.end_datapipe = None
```
Comment (ejguan) [inline torchdata/dataloader2/reading_service.py:191]: I change it to `end_datapipe` because we need to store the last DataPipe for both In-process and multiprocessing cases.

File: torchdata/dataloader2/reading_service.py [hunk 3]
```diff
@@ -215,12 +263,33 @@ def initialize(self, datapipe: DataPipe) -> DataPipe:
...
+        shared_seed = shared_seed.item()
+        _seed_generator = torch.Generator()
+        _seed_generator.manual_seed(shared_seed)
+        torch.utils.data.graph_settings.apply_random_seed(
+            self.end_datapipe,
+            _seed_generator,
+        )
+
+        # Multiprocessing (num_workers > 0)
+        if isinstance(self.end_datapipe, _IterateQueueDataPipes):
```
Comment (VitalyFedyunin) [inline torchdata/dataloader2/reading_service.py:283]: Will merge conflict with my prefetcher fix

File: torchdata/dataloader2/reading_service.py [hunk 4]
```diff
@@ -174,38 +188,72 @@ def __init__(
...
+        worker_seed_generator.manual_seed(shared_seed)
+        torch.utils.data.graph_settings.apply_random_seed(
+            datapipe,
+            worker_seed_generator,
+        )
+        # Set different seeds across distributed workers
+        global_worker_id = worker_id * world_size + rank
+        worker_seed_generator.manual_seed(shared_seed + global_worker_id)
+
+        import random
```
Comment (VitalyFedyunin) [inline torchdata/dataloader2/reading_service.py:215]: Any reason to import inside of the function?

File: torchdata/dataloader2/reading_service.py [hunk 5]
```diff
@@ -174,38 +188,72 @@ def __init__(
...
-        # TODO(615): Add shuffle determinism support
-        torch.utils.data.graph_settings.apply_sharding(datapipe, num_workers, worker_id)
+    def _process_init_fn(world_size, rank, num_workers, worker_id, datapipe):
+        global_worker_id = worker_id * world_size + rank
+        total_num_workers = num_workers * world_size
+        torch.utils.data.graph_settings.apply_sharding(datapipe, total_num_workers, global_worker_id)
 
     @staticmethod
-    def call_on_epoch_reset(datapipe, *args):
+    def _process_reset_fn(world_size, rank, num_workers, worker_id, datapipe, shared_seed, *args):
```
Comment (VitalyFedyunin) [inline torchdata/dataloader2/reading_service.py:203]: Feel free to drop `*args` here, as you accept just datapipe and seed.

Comment (ejguan) [reply torchdata/dataloader2/reading_service.py:203]: For sure. And, those APIs and messages are private, so I think it's totally fine to remove.

Comment (NivekT) [reply torchdata/dataloader2/reading_service.py:203]: nit: `num_workers` also isn't needed here

**Review by NivekT — APPROVED** [review]

File: torchdata/dataloader2/reading_service.py [hunk 6]
```diff
@@ -174,38 +188,72 @@ def __init__(
         self.multiprocessing_context = multiprocessing_context
         self.processes = []
         self.datapipes = []
-        self.combined_datapipes = None
+        self.end_datapipe = None
+        self._pg = None
+        self._world_size = 1
+        self._rank = 0
```
[the remark this answers is not in the document]
Comment (ejguan) [reply torchdata/dataloader2/reading_service.py:195]: Yeah, that's what I expected.

File: torchdata/dataloader2/reading_service.py (hunk 5 above)
Comment (ejguan) [reply torchdata/dataloader2/reading_service.py]: > nit: `num_workers` also isn't needed here

Yeah, I intentionally add it here to align with those two static functions for now. With `SequentialRS`, we only need `worker_id`, `datapipe` and a `seed_generator` here.

**Review by VitalyFedyunin — COMMENTED** [review]
LGTM

**Review by VitalyFedyunin — APPROVED** [review]

**Review by NivekT — APPROVED** [review]
